firedrake/assign.py

```diff
@@ -0,0 +1,130 @@
+        self.weighted_coefficients.append((weight, coeff))
+
+    def coefficient(self, o):
+        self.weighted_coefficients.append((1, o))
+
+    def invalid(self, o):
+        raise ValueError(f"{o} is not a valid component of a weighted sum of coefficients")
+
+    expr = invalid
+
+
+class Assigner:
+    symbol = "="
+
+    def __init__(self, assignee, expression, subset=None):
+        if isinstance(expression, Vector):
+            expression = expression.function
+
+        if not all(
+            c.function_space() == assignee.function_space()
+            for c in extract_coefficients(expression)
+        ):
+            raise ValueError
+
+        self._assignee = assignee
+        self._expression = expression
+        self._subset = subset
+
+    def __str__(self):
+        return f"{self._assignee} {self.symbol} {self._expression}"
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._assignee!r}, {self._expression!r})"
+
+    def assign(self):
+        if self._subset:
+            indices = self._subset.indices
+            self._assignee.dat.data[indices] = np.dot(self.weights, self.coefficient_data[indices])
+        else:
+            # if there is no subset then we can also assign halo values to avoid a later halo exchange
+            self._assignee.dat.data_with_halos[...] = np.dot(self.weights, self.coefficient_data_with_halos)
+            self._assignee.dat.halo_valid = True
+
+    @cached_property
+    def weights(self):
+        return np.array(self._weighted_coefficients[0], dtype=ScalarType)
+
+    @cached_property
+    def coefficients(self):
+        return self._weighted_coefficients[1]
+
+    @cached_property
+    def coefficient_data(self):
+        return np.array([c.dat.data_ro for c in self.coefficients], dtype=ScalarType)
+
+    @cached_property
+    def coefficient_data_with_halos(self):
+        return np.array([c.dat.data_ro_with_halos for c in self.coefficients], dtype=ScalarType)
+
+    @cached_property
+    def _weighted_coefficients(self):
+        # TODO: It would be nice to stash the weights and coefficient vecs on the expression
+        # but expressions do not currently have caches attached to them.
+        collector = WeightedCoefficientCollector()
+        map_expr_dag(collector, self._expression)
+        return tuple(zip(*collector.weighted_coefficients))
+
+
+class AugmentedAssigner(Assigner, abc.ABC):
+    @cached_property
+    def coefficient_vecs(self):
+        assert not self._subset
+        return tuple(c.dat._vec for c in self.coefficients)
+
+
+class IAddAssigner(AugmentedAssigner):
+    symbol = "+="
+
+    def assign(self):
+        if self._subset:
+            indices = self._subset.indices
+            self._assignee.dat.data[indices] += np.dot(self.weights, self.coefficient_data[indices])
+        else:
+            # TODO Explain why maxpy preferred to numpy
+            self._assignee.dat._vec.maxpy(self.weights, self.coefficient_vecs)
```

Comment:
Careful, this is unsafe in the case that the lvalue or rvalues are mixed dats (the `_vec` is not a view into the data in that case)